ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+
+module Test.Ouroboros.Storage.ChainDB.Model (
+    tests
+  ) where
+
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import           Ouroboros.Consensus.Protocol.Abstract
+import           Ouroboros.Network.Block (HasHeader (..))
+import qualified Ouroboros.Storage.ChainDB.Model as M
+import           Test.Ouroboros.Storage.ChainDB.TestBlock
+
+tests :: TestTree
+tests = testGroup "Model" [
+      testProperty "getBlock_addBlock"        prop_getBlock_addBlock
+    , testProperty "getChain_addChain"        prop_getChain_addChain
+    , testProperty "alwaysPickPreferredChain" prop_alwaysPickPreferredChain
+    ]
+
+prop_getBlock_addBlock :: BlockTree -> Property
+prop_getBlock_addBlock bt =
+        M.getBlock (blockHash lastBlock) (M.addBlock testConfig lastBlock model)
+    === Just lastBlock
+  where
+    blocks     = treeToBlocks bt
+    initBlocks = init blocks
+    lastBlock  = last blocks
```

Comment:
Is it significant which block the `lastBlock` is? Why not pattern-match on `blocks` to extract the head and the tail? Or does the order of the list returned by `treeToBlocks` matter? In that case it should be written down somewhere.

Reply:
Well, the order returned by `treeToBlocks` is an in-order traversal. The first block will always be the first block on the chain. But you are right nonetheless; I'll reuse the same permutation infrastructure to permute the list of blocks and then just pattern match and get the (now random) first one as the "new" one.